src/general/DayNightCycle.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using Newtonsoft.Json;
+
+[JsonObject(IsReference = true)]
+[UseThriveSerializer]
+public class DayNightCycle
+{
+    [JsonProperty]
+    public DayNightConfiguration LightCycleConfig;
+
+    [JsonProperty]
+    public bool IsEnabled;
+
+    /// <summary>
+    ///   The multiplier used for calculating DayLightPercentage.
+    /// </summary>
+    /// <remarks>
+    ///   This exists as it only needs to be calculated once and
+    ///   the calculation for it is confusing.
+    /// </remarks>
+    [JsonIgnore]
+    private float daytimeMultiplier;
+
+    public DayNightCycle(bool isEnabled)
+    {
+        IsEnabled = isEnabled;
+
+        LightCycleConfig = SimulationParameters.Instance.GetDayNightCycleConfiguration();
+        Time = LightCycleConfig.HoursPerDay / 2;
+
+        float halfPercentage = LightCycleConfig.DaytimePercentage / 2;
+
+        // This converts the percentage in DaytimePercentage to the power of two needed for DayLightPercentage
+        daytimeMultiplier = (float)Math.Pow(2, 1 / halfPercentage);
+
+        if (IsEnabled)
+        {
+            AverageSunlight = EvaluateAverageSunlight(0.5f + halfPercentage)
+                - EvaluateAverageSunlight(0.5f - halfPercentage);
+        }
+        else
+        {
+            AverageSunlight = 1;
+        }
+    }
+
+    [JsonIgnore]
+    public float AverageSunlight { get; private set; }
+
+    /// <summary>
+    ///   The current time in hours
+    /// </summary>
+    [JsonProperty]
+    public float Time { get; private set; }
+
+    [JsonIgnore]
+    public float PercentOfDayElapsed => Time / LightCycleConfig.HoursPerDay;
+
+    /// <summary>
+    ///   The percentage of daylight you should get.
+    ///   light = max(-(PercentOfDayElapsed - 0.5)^2 * daytimeMultiplier + 1, 0)
+    ///   desmos: https://www.desmos.com/calculator/vrrk1bkac2
+    /// </summary>
+    /// <remarks>
+    ///   If this equation is changed EvaluateAverageSunlight needs to be updated.
+    /// </remarks>
+    [JsonIgnore]
+    public float DayLightPercentage =>
+        Math.Max(-(float)Math.Pow(PercentOfDayElapsed - 0.5, 2) * daytimeMultiplier + 1, 0);
+
+    public void Process(float delta)
+    {
+        if (IsEnabled)
+        {
+            Time = (Time + (1 / LightCycleConfig.RealTimePerDay) * LightCycleConfig.HoursPerDay * delta)
+                % LightCycleConfig.HoursPerDay;
```

Comment:
```suggestion
            Time = (Time + delta  * LightCycleConfig.HoursPerDay / LightCycleConfig.RealTimePerDay)
                % LightCycleConfig.HoursPerDay;
```

But also see my suggestion for not using absolute time.